### PR TITLE
TST: machinery for tests requiring large memory + lapack64 smoketest

### PR DESCRIPTION
**Pull request by pv**
Implement test suite machinery for (automatically) skipping tests that require large amounts of memory.

Also, add a companion test for gh-15012, which checks 64-bit LAPACK works as expected.

(I ended up adding `np.dot` check instead of the `eigh` check, because although the `eigh` one seems to take less memory, the runtime seems to be > 20min so it's not suitable here.)

### CODE REVIEW

**Comment by pv** [comment]
Azure failures seems spurious ("[NuGet] The operation has timed out" when installing mingw)

**Comment by charris** [comment]
Thanks Pauli.

File: numpy/lib/tests/test_io.py [hunk 1]
```diff
@@ -575,13 +576,9 @@ def test_unicode_and_bytes_fmt(self, fmt, iotype):
     @pytest.mark.skipif(sys.platform=='win32',
                         reason="large files cause problems")
     @pytest.mark.slow
+    @requires_memory(7e9)
```
Comment (eric-wieser) [inline numpy/lib/tests/test_io.py]: nit - the original spelling seemed less magic, and had the bonus of being integral
```suggestion
    @requires_memory(6 * 1024 * 1024 * 1024)
```

File: numpy/testing/_private/utils.py [hunk 2]
```diff
@@ -2380,3 +2382,73 @@ def break_cycles():
...
+        return psutil.virtual_memory().available
+    except (ImportError, AttributeError):
+        pass
+
+    if sys.platform.startswith('linux'):
+        info = {}
+        with open('/proc/meminfo', 'r') as f:
+            for line in f:
+                p = line.split()
+                info[p[0].strip(':').lower()] = float(p[1]) * 1e3
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]:
```suggestion
                info[p[0].strip(':').lower()] = int(p[1]) * 1024
```
Based on https://lore.kernel.org/patchwork/patch/666444/, this measurement is in KiB

File: numpy/testing/_private/utils.py [hunk 3]
```diff
@@ -2380,3 +2382,73 @@ def break_cycles():
...
+                free_bytes/1e9, mem_free/1e9)
+
+    return pytest.mark.skipif(mem_free < free_bytes, reason=msg)
+
+
+def _parse_size(size_str):
+    """Convert memory size strings ('12 GB' etc.) to float"""
+    suffixes = {'': 1.0, 'b': 1.0,
+                'k': 1e3, 'm': 1e6, 'g': 1e9, 't': 1e12,
+                'kb': 1e3, 'mb': 1e6, 'gb': 1e9, 'tb': 1e12}
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: nit: consider including `kib`, `MiB`, etc

File: numpy/testing/_private/utils.py [hunk 4]
```diff
@@ -2380,3 +2382,73 @@ def break_cycles():
...
+        else:
+            msg = '{0} GB memory required, but {1} GB available'.format(
+                free_bytes/1e9, mem_free/1e9)
+
+    return pytest.mark.skipif(mem_free < free_bytes, reason=msg)
+
+
+def _parse_size(size_str):
+    """Convert memory size strings ('12 GB' etc.) to float"""
+    suffixes = {'': 1.0, 'b': 1.0,
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: nit: Perhaps for brevity:
```
suffixes = {'': 1.0, 'k': 1e3, 'm': 1e6, 'g': 1e9, 't': 1e12}
suffixes.update({k + 'b': v for k, v in suffixes.items()})  # allow a trailing b

File: numpy/testing/_private/utils.py [hunk 5]
```diff
@@ -2380,3 +2382,73 @@ def break_cycles():
...
+        mem_free = _get_mem_available()
+
+        if mem_free is None:
+            msg = ("Could not determine available memory; set NPY_AVAILABLE_MEM "
+                   "environment variable (e.g. NPY_AVAILABLE_MEM=16GB) to run "
+                   "the test.")
+            mem_free = -1
+        else:
+            msg = '{0} GB memory required, but {1} GB available'.format(
+                free_bytes/1e9, mem_free/1e9)
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: nit: could use [`_ArrayMemoryError._size_to_string`](https://github.com/numpy/numpy/blob/master/numpy/core/_exceptions.py#L158) here to format numbers as prefixed bytes.

File: numpy/testing/_private/utils.py [hunk 6]
```diff
@@ -2380,3 +2382,73 @@ def break_cycles():
...
+    env_value = os.environ.get(env_var)
+    if env_value is not None:
+        try:
+            mem_free = _parse_size(env_value)
+        except ValueError as exc:
+            raise ValueError('Invalid environment variable {}: {!s}'.format(
+                env_var, exc))
+
+        msg = ('{0} GB memory required, but environment variable '
+               'NPY_AVAILABLE_MEM={1} set'.format(
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]:
```suggestion
               'NPY_AVAILABLE_MEM="{1} GB" set'.format(
```

File: numpy/testing/_private/utils.py [hunk 7]
```diff
@@ -2380,3 +2382,73 @@ def break_cycles():
...
+                'k': 1e3, 'm': 1e6, 'g': 1e9, 't': 1e12,
+                'kb': 1e3, 'mb': 1e6, 'gb': 1e9, 'tb': 1e12}
+
+    size_re = re.compile(r'^\s*(\d+|\d+\.\d+)\s*({0})\s*$'.format(
+        '|'.join(suffixes.keys())), re.I)
+
+    m = size_re.match(size_str.lower())
+    if not m or m.group(2) not in suffixes:
+        raise ValueError("value {!r} not a valid size".format(size_str))
+    return float(m.group(1)) * suffixes[m.group(2)]
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: nit: sizes are ints,
```suggestion
    return int(float(m.group(1)) * suffixes[m.group(2)])
```

File: numpy/testing/_private/decorators.py [hunk 8]
```diff
@@ -15,6 +15,10 @@
 """
 from __future__ import division, absolute_import, print_function
 
+import sys
+import os
+import re
+
```
Comment (eric-wieser) [inline numpy/testing/_private/decorators.py]: Why this change?

**Review by eric-wieser — COMMENTED** [review]
Mostly looks good, one bug and some nits

**Comment by eric-wieser** [comment]
@charris, looks like you merged while I was reviewing. It's unfortunate github doesn't provide an easy way to collect the diffs above into a patch once it's merged.

**Comment by charris** [comment]
@eric-wieser The changes look small, just make a new PR. I was kinda counting on that anyway :)

File: numpy/testing/_private/utils.py (hunk 2 above)
Comment (pv) [reply numpy/testing/_private/utils.py]: Ok, thanks. (follow-up in gh-15052)

File: numpy/testing/_private/utils.py (hunk 3 above)
Comment (pv) [reply numpy/testing/_private/utils.py]: Added

File: numpy/testing/_private/utils.py (hunk 4 above)
Comment (pv) [reply numpy/testing/_private/utils.py]: After adding kib, it seemed cleaner to not do this.

File: numpy/testing/_private/utils.py (hunk 6 above)
Comment (pv) [reply numpy/testing/_private/utils.py]: {1} is the string value of the env var here

File: numpy/testing/_private/utils.py (hunk 7 above)
Comment (pv) [reply numpy/testing/_private/utils.py]: ok

File: numpy/testing/_private/decorators.py (hunk 8 above)
Comment (pv) [reply numpy/testing/_private/decorators.py]: Added by mistake.

File: numpy/testing/_private/utils.py (hunk 5 above)
Comment (pv) [reply numpy/testing/_private/utils.py]: Didn't change this: corner-case message in test suite, and probably would have involved refactoring the utility function out from `_ArrayMemoryError`.

Comment (eric-wieser) [reply numpy/testing/_private/utils.py]: Fair
